### PR TITLE
Add support for Razer Naga X

### PR DESCRIPTION
**Pull request by daenney**
I've implemented this based on what was discussed in #1463. I'm still waiting on the mouse to arrive to check everything, but I'm hoping people might be willing to test this beforehand.

The device seems almost identical to the Naga Left Handed, so I've based it on that. The main difference seems to be the Naga Left Handed has 2 buttons under the scroll wheel, whereas the Naga X has one. Idk if this requires any specific handling? The button on the Naga X is known as the "DPI cycle button", whereas the Naga Left Handed has "DPI Up" and "DPI down" buttons.

~~I'm not a 100% clear on if `dev_attr_dpi_stages` is actually accurate/available for this device, but the support docs/tech specs are worded in a similar way as other devices for which this is enabled.~~ It does support DPI stages based in the pcap from #1463

### CODE REVIEW

**Comment by daenney** [comment]
@z3ntu In #1463 you mentioned

> `razer_chroma_extended_matrix_effect_*` effects with ZERO_LED

I took that as all `razer_chroma_extended_matrix_effect_*` functions should take `ZERO_LED`, but it strikes me as somewhat odd that `razer_attr_write_scroll_mode_breath` for example which calls `razer_chroma_extended_matrix_effect_breathing_*` etc. would take `ZERO_LED` instead of `SCROLL_WHEEL_LED`?

**Comment by Vvamp** [comment]
@daenney I've got a razer naga x and would love to help you test it, if you still need it. I cloned and installed your fork. How can I help you test it?

**Comment by daenney** [comment]
@Vvamp Try to use one of the apps, see if the mouse shows up, if you can configure the lighting effects etc.

**Comment by daenney** [comment]
My Naga X arrives this week so I'll hopefully have some time this weekend to validate everything myself.

**Comment by Vvamp** [comment]
Alright, I used polychromatic and razercommander. Unfortunately it seems like it is not recognizing the naga x, at least for me. 
I might have done something wrong, ofcourse, but I restarted the daemon and I shouldn't have any other versions of openrazer installed. 
Using `cat device_type` directly does however give back that it's the razer naga x.

**Comment by daenney** [comment]
Good to know. It's very possible I've missed wiring something up, I'm still very new to this code base and it's a bit hard to debug this remotely. I'll try to have a go at it this weekend time permitting.

**Comment by daenney** [comment]
Alright, so I'm running into an issue adding the device itself:

```
Nov 26 11:53:52 monoceros openrazer-daemon[1448]: 2021-11-26 11:53:52 | razer                          | INFO     | Found valid device.0: 0003:1532:0096.0012
Nov 26 11:53:52 monoceros openrazer-daemon[1448]: 2021-11-26 11:53:52 | razer.device0                  | INFO     | Initialising device.0 RazerNagaX
Nov 26 11:53:57 monoceros openrazer-daemon[1448]: 2021-11-26 11:53:57 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xd>
Nov 26 11:54:02 monoceros openrazer-daemon[1448]: Exception in thread Thread-2:
Nov 26 11:54:02 monoceros openrazer-daemon[1448]: Traceback (most recent call last):
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/threading.py", line 973, in _bootstrap_inner
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     self.run()
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/threading.py", line 910, in run
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     self._target(*self._args, **self._kwargs)
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/daemon.py", line 597, in _collecting_udev_method
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     self._add_device(d)
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/daemon.py", line 522, in _add_device
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     razer_device = device_class(device_path=sys_path, device_number=device_number, config=self._config,
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/hardware/device_base.py", line 119, in __init__
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     super().__init__(object_path)
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/dbus_services/service.py", line 55, in __init__
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     self.add_to_connection(bus, object_path)
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:   File "/usr/lib/python3.9/site-packages/dbus/service.py", line 582, in add_to_connection
Nov 26 11:54:02 monoceros openrazer-daemon[1448]:     connection._register_object_path(path, self._message_cb,
Nov 26 11:54:02 monoceros openrazer-daemon[1448]: ValueError: Invalid object path '/org/razer/device/@I': contains invalid character '@'
```

`WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xd>` would seem to be an issue, but I have no idea what prompts this.

**Comment by daenney** [comment]
Seems like this error doesn't always happen, which suggest a timing issue?

When it gets past it, I get to some other trouble:

```
Nov 26 12:09:23 monoceros openrazer-daemon[5122]: Exception in thread Thread-2:
Nov 26 12:09:23 monoceros openrazer-daemon[5122]: Traceback (most recent call last):
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:   File "/usr/lib/python3.9/threading.py", line 973, in _bootstrap_inner
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:     self.run()
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:   File "/usr/lib/python3.9/threading.py", line 910, in run
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:     self._target(*self._args, **self._kwargs)
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/daemon.py", line 598, in _collecting_udev_method
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:     self._add_device(d)
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/daemon.py", line 522, in _add_device
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:     razer_device = device_class(device_path=sys_path, device_number=device_number, config=self._config,
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/hardware/device_base.py", line 316, in __init__
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:     bright_func(self.zone[i]["brightness"])
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/dbus_services/dbus_methods/lanceheadte.py", line 92, in set_left_brightness
Nov 26 12:09:23 monoceros openrazer-daemon[5122]:     with open(driver_path, 'w') as driver_file:
Nov 26 12:09:23 monoceros openrazer-daemon[5122]: PermissionError: [Errno 13] Permission denied: '/sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-6/3-6:1.0/0003:1532:0096.0016/left_led_brightness'
```

Now this isn't really a permission error, that file doesn't exist:
```
ls -hFGa /sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-6/3-6:1.0/0003:1532:0096.0016
./             dpi_stages                 logo_matrix_effect_none      modalias                    right_matrix_effect_reactive  scroll_matrix_effect_reactive  version
../            driver@                    logo_matrix_effect_reactive  poll_rate                   right_matrix_effect_spectrum  scroll_matrix_effect_spectrum
country        firmware_version           logo_matrix_effect_spectrum  power/                      right_matrix_effect_static    scroll_matrix_effect_static
device_mode    hidraw/                    logo_matrix_effect_static    report_descriptor           right_matrix_effect_wave      scroll_matrix_effect_wave
device_serial  input/                     logo_matrix_effect_wave      right_led_brightness        scroll_led_brightness         subsystem@
device_type    logo_led_brightness        matrix_custom_frame          right_matrix_effect_breath  scroll_matrix_effect_breath   test
dpi            logo_matrix_effect_breath  matrix_effect_custom         right_matrix_effect_none    scroll_matrix_effect_none     uevent
```

So I think I wired something up wrong. It should be left, not right.

**Comment by daenney** [comment]
> So I think I wired something up wrong. It should be left, not right.

Indeed I did. I based it off the Naga Left Handed and forgot to remap the attributes from right to left in `razer_mouse_probe`.

**Comment by Vvamp** [comment]
Let me know if if works now, or if there are any other issues! I can't test it myself rn, but I will be later today. I might also be able to help with some bugs, but I'm not very experienced in device drivers

**Comment by daenney** [comment]
Alright, so it works for me now in that the device is mostly recognised and it shows up in Polychromatic. However, it all feels very buggy currently. Polychromatic feels like it hangs constantly and though I managed to turn the logo effect off I can't get it to go back on.

**Comment by Vvamp** [comment]
I'll test in a couple of hours and let you know how it works for me

**Comment by daenney** [comment]
I think I might be confusing something. The Naga X doesn't have a logo led, does it?

**Comment by Vvamp** [comment]
Don't think so iirc, only scroll wheel and button area

**Comment by daenney** [comment]
Cool, thanks for confirming. For a second there I thought I broke it :sweat_smile:

**Comment by daenney** [comment]
So from where I'm standing the mouse is recognised but changing the effects or brightness of the scroll wheel or thumb cluster isn't doing anything. I also don't think it's reading the brightness attributes correctly.

**Comment by daenney** [comment]
I'm still trying to figure out why reading the device serial is not working. I tried upping the timeout for reading those attributes but that doesn't seem to do the trick.

All I get is this:
```
>>> f=open("/sys/devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-6/3-6:1.0/0003:1532:0096.003A/device_serial", "rb")
>>> print(f.read())
b'/\x01\n'
```
That would just be `1`, which seems wrong. I also have cases where it returns `0xdd`, i.e 221 but that doesn't seem to make much sense either.

**Comment by daenney** [comment]
I also still occasionally get this, but I suspect that's due to not properly reading some attributes like the device serial.

```
Nov 26 14:13:27 monoceros openrazer-daemon[125639]: Traceback (most recent call last):
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/lib/python3.9/site-packages/daemonize.py", line 248, in start
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     self.action(*privileged_action_result)
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/bin/openrazer-daemon", line 140, in run_daemon
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     daemon = RazerDaemon(verbose=args.verbose,
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/daemon.py", line 115, in __init__
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     self._load_devices(first_run=True)
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/daemon.py", line 483, in _load_devices
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     razer_device = device_class(device_path=sys_path, device_number=device_number, config=self._config,
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/hardware/device_base.py", line 119, in __init__
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     super().__init__(object_path)
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/lib/python3.9/site-packages/openrazer_daemon/dbus_services/service.py", line 55, in __init__
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     self.add_to_connection(bus, object_path)
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:   File "/usr/lib/python3.9/site-packages/dbus/service.py", line 582, in add_to_connection
Nov 26 14:13:27 monoceros openrazer-daemon[125639]:     connection._register_object_path(path, self._message_cb,
Nov 26 14:13:27 monoceros openrazer-daemon[125639]: [68B blob data]
```

**Comment by Vvamp** [comment]
alright so I tested it at home again. Uninstalled previous versions, re-cloned your repo, switched to nagax branch and installed with 'make install_i_know_what_i_am_doing'. 

Once it was done, rebooted and checked polychromatic. For me, it's still not showing up, oddly  enough

**Comment by daenney** [comment]
I think you need to update your udev rules. If you look in my PR you'll see it. I'm rebuilding and installing it based on the AUR package which ensures those rules are in place.

**Comment by daenney** [comment]
Alternatively, check the log output of the openrazer-daemon.

**Comment by Vvamp** [comment]
I believe `make install_i_know_what_i_am_doing` installs the udev rulesfor me.

 However, I did check the logs. For me, it today stopped showing any devices on my razer and it seems to be related to that warning you were getting. When tailing the log file and restarting the daemon, polychromatic  seems to continue waiting for the backend to start, all the while the log spitting the warning a few times. 

After about 8 warnings or so, it stops and the backend does start, but without any devices being recognized.
This is all I'm getting: 

> 2021-11-27 21:29:37 | razer                          | INFO     | Initialising Daemon (v3.1.0). Pid: 68979
> 2021-11-27 21:29:37 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
> 2021-11-27 21:29:37 | razer                          | INFO     | Found device.0: 0003:1532:0096.0007
> 2021-11-27 21:29:37 | razer.device0                  | INFO     | Initialising device.0 RazerNagaX
> 2021-11-27 21:29:47 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:29:52 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:30:03 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:30:18 | razer                          | INFO     | Initialising Daemon (v3.1.0). Pid: 79439
> 2021-11-27 21:30:18 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
> 2021-11-27 21:30:18 | razer                          | INFO     | Found device.0: 0003:1532:0096.0007
> 2021-11-27 21:30:18 | razer.device0                  | INFO     | Initialising device.0 RazerNagaX
> 2021-11-27 21:30:24 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:30:29 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:30:34 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:30:39 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte
> 2021-11-27 21:30:44 | razer.device0                  | WARNING  | malformed serial: 'utf-8' codec can't decode byte 0xdd in position 0: invalid continuation byte

**Comment by daenney** [comment]
Right, yeah. So idk what's up with that. I'd really need a maintainer who might have seen this before to chime in at this point. As far as I can tell everything's wired up sensibly, but clearly something's off.

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 1]
```diff
@@ -1981,6 +1981,65 @@ def _resume_device(self):
...
+        _set_left_brightness(self, 0)
+        self.disable_notify = False
+
+    def _resume_device(self):
+        """
+        Resume the device
+        Get the last known brightness and then set the brightness
+        """
+        scroll_brightness = self.suspend_args.get('brightness', (100, 100, 100))[0]
+        left_row_brightness = self.suspend_args.get('brightness', (100, 100, 100))[1]
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: The fallback parameter should be `(100, 100)`, not `(100, 100, 100)` as you're only ever passing in two values in `_suspend_device`

**Comment by z3ntu** [comment]
Given that reading the serial number doesn't work properly, can you check dmesg if there are any errors from razermouse? Also you can try using a different transaction id in `razer_attr_read_get_serial`, maybe that changes something.

**Comment by daenney** [comment]
dmesg gives me this right now when I plug in the mouse:

```
[139181.425094] usb 3-6: new full-speed USB device number 4 using xhci_hcd
[139181.590769] usb 3-6: New USB device found, idVendor=1532, idProduct=0096, bcdDevice= 2.00
[139181.590774] usb 3-6: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[139181.590776] usb 3-6: Product: Razer Naga X
[139181.590777] usb 3-6: Manufacturer: Razer
[139181.616911] input: Razer Razer Naga X as /devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-6/3-6:1.0/0003:1532:0096.0038/input/input71
[139181.616991] razermouse 0003:1532:0096.0038: input,hidraw2: USB HID v1.00 Mouse [Razer Razer Naga X] on usb-0000:07:00.3-6/input0
[139181.621958] input: Razer Razer Naga X as /devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-6/3-6:1.1/0003:1532:0096.0039/input/input72
[139181.678538] razermouse 0003:1532:0096.0039: input,hidraw3: USB HID v1.00 Keyboard [Razer Razer Naga X] on usb-0000:07:00.3-6/input1
[139181.680937] input: Razer Razer Naga X as /devices/pci0000:00/0000:00:01.2/0000:02:00.0/0000:03:08.0/0000:07:00.3/usb3/3-6/3-6:1.2/0003:1532:0096.003A/input/input73
[139181.735278] razermouse 0003:1532:0096.003A: input,hidraw4: USB HID v1.00 Keyboard [Razer Razer Naga X] on usb-0000:07:00.3-6/input2
[139181.737850] razermouse 0003:1532:0096.003B: No inputs registered, leaving
[139181.737869] razermouse 0003:1532:0096.003B: hidraw5: USB HID v1.00 Device [Razer Razer Naga X] on usb-0000:07:00.3-6/input3
```

**Comment by daenney** [comment]
@z3ntu I tried with both other branches in `razer_attr_read_get_serial` but same result, I keep getting that malformed serial.

The device does show up in Polychromatic for me with the UI elements I expect, but nothing actually works. Changing any setting is sluggish / feels like it hangs the whole app and doesn't result in any change to the device itself (like changing the light effect or level).

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 1 above)
Comment (daenney) [reply daemon/openrazer_daemon/hardware/mouse.py]: Thanks for catching this, fixed now.

**Comment by z3ntu** [comment]
Weird. For now you can try to hack the daemon to use a random serial and ignore that this is not working
```
daemon/openrazer_daemon/hardware/device_base.py:                serial = 'UNKWN{0:012}'.format(random.randint(0, 4096))
```
Given it's a fairly recent device, there's a chance that Razer changed something in the firmware and doesn't return the serial number as we expect. Or Razer messed up while programming the serial number into the device, we've also seen that before.

**Comment by daenney** [comment]
So I'm wondering if something more fundamental might have changed. In Synapse I can only control scroll wheel and thumbpad at the same time, I have to drop into Chroma to do them individually.

Capturing data from Synapse, I see stuff like:
```
0000   00 ef ef c9 ae 97 ff ff 53 02 00 04 03 00 00 00
0010   b2 f2 ab 61 00 00 00 00 b0 b4 0e 00 8d ff ff ff
0020   5a 00 00 00 5a 00 00 00 21 09 00 03 03 00 5a 00
0030   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

0040   00 1f 00 00 00 0b 0f 03 00 00 00 00 01 00 ff 00
0050   00 ff 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0060   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0090   00 00 00 00 00 00 00 00 06 00
```

I'm recognising part of the packet, starting on `0040` with `00 1f`, then `00 00 00` (which is the reserved bits?), followed by `0b`, 11(?) for parameters, `0f` reserved, `03` command, `00 00 00 00 01` sub command (this can't be right?) and then `00 ff 00` twice, setting the scroll wheel and the thumpad to green.

I feel like I'm misunderstanding something, but it doesn't seem to line up quite right with what's described in https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol#phase-4---decoding-the-protocol.

**Comment by z3ntu** [comment]
The `0f 03` part is exactly this command here: https://github.com/openrazer/openrazer/blob/master/driver/razerchromacommon.c#L753 You can also see the parameters there with names.

**Comment by daenney** [comment]
Aha. It seems a lot of those functions have `report.transaction_id.id = 0x3F` hardcoded into them. Based on what I'm seeing I need `0x1F`?

**Comment by z3ntu** [comment]
The user of these functions should set the transaction id correctly (so in your case where e.g. `razer_chroma_extended_matrix_set_custom_frame` is being used in `razermouse_driver.c`)

**Comment by ProfOak** [comment]
Hey all, I discovered something that I couldn't find covered elsewhere on the internet so I thought I'd share it here. It's not exactly related to this PR but I hope it can be helpful for future work.

I recently bought a Razer Naga X and discovered you can set the keys with the software in Windows, with the official Razer Synapse program, and have the key configurations carry over to Linux. I noticed this after dual booting from Windows to Linux but it might be possible to configure the mouse on a Windows VM if you pass through the USB device. You may need a Razer account to use the Synapse software. Other than that it's a relatively simple way to set key configurations and possibly other settings for the mouse on Linux.

**Comment by daenney** [comment]
I unfortunately don't have time to work on this any more. Someone else is welcome to take the changes and see if they can make it work though.

**Comment by bufothefrog** [comment]
I'm putting a $20 USD bounty on this getting solved, payable via Zelle/PayPal, charitable donation or XMR. Feel free to remove if this isn't allowed.

**Comment by bourquenoud** [comment]
I've recently bought a Naga X and realised  it wasn't supported by openrazer. So now I'm working on this PR.

The first issue is that `razer_attr_read_get_serial` won't get the correct response from the mouse. I took a look at the packets sent by Synapse and apparently the transaction ID is neither `0x3F` nor `0x1F`, but `0x08`.

I modified the driver to use the correct transaction ID, but the mouse still won't answer with a valid packet even though the command sent is identical.

If someone more experienced has an explanation, I'd gladly hear it...

I'm willing to implement the support for the Naga X but this issue doesn't make any sense...

---

With Windows VM  + Synapse :
```
0000   40 3e cd fd 54 95 ff ff 53 02 00 02 01 00 00 00
0010   d5 ca ef 62 00 00 00 00 47 92 0e 00 8d ff ff ff
0020   5a 00 00 00 5a 00 00 00 21 09 00 03 03 00 5a 00
0030   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

0040   00 08 00 00 00 16 00 82 00 00 00 00 00 00 00 00
0050   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0060   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0090   00 00 00 00 00 00 00 00 94 00
```

answer :
```
0000   00 3c cd fd 54 95 ff ff 43 02 80 02 01 00 2d 00
0010   d5 ca ef 62 00 00 00 00 e1 cf 0e 00 00 00 00 00
0020   5a 00 00 00 5a 00 00 00 00 00 00 00 00 00 00 00
0030   00 00 00 00 00 00 00 00 00 02 00 00 00 00 00 00

0040   02 08 00 00 00 16 00 82 50 4d 32 31 30 33 48 31
0050   38 30 30 35 32 37 36 00 00 00 00 00 00 00 00 00
0060   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0090   00 00 00 00 00 00 00 00 ce 00
```

Sent by openrazer :
```
0000   00 f0 69 6e 68 91 ff ff 53 02 00 15 01 00 00 00
0010   3e d9 ef 62 00 00 00 00 50 2f 0b 00 8d ff ff ff
0020   5a 00 00 00 5a 00 00 00 21 09 00 03 00 00 5a 00
0030   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

0040   00 08 00 00 00 16 00 82 00 00 00 00 00 00 00 00
0050   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0060   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0070   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0080   00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0090   00 00 00 00 00 00 00 00 94 00
```

answer : 
```
0000   80 4a bd 7c 68 91 ff ff 43 02 80 15 01 00 2d 00
0010   43 d9 ef 62 00 00 00 00 bf 03 0d 00 00 00 00 00
0020   5a 00 00 00 5a 00 00 00 00 00 00 00 00 00 00 00
0030   00 00 00 00 00 00 00 00 00 02 00 00 00 00 00 00

0040   fc 7f 00 04 d5 00 00 00 dd 00 00 00 df 00 00 00
0050   00 00 00 00 00 00 00 00 00 00 00 00 73 7d ff fb
0060   00 00 00 00 00 00 00 00 00 00 00 00 e1 00 00 00
0070   00 00 00 00 00 00 00 00 e3 00 00 00 e5 00 00 00
0080   e7 00 00 00 eb 00 00 00 ef 00 00 00 f3 00 00 00
0090   f7 00 00 00 fb 00 00 00 ff 00
```

**Comment by bourquenoud** [comment]
It is now working, I've made a PR (#1878).

@bufothefrog, you now owe me 20 bucks :wink:. But the maintainers of OpenRazer would make better use of this money, so give it to them instead (https://www.paypal.me/z3ntu, https://www.paypal.me/LukeHorwell and https://www.paypal.me/terrycain1337)

**Comment by bufothefrog** [comment]
@bourquenoud https://imgur.com/a/r1zMWZO. I was too lazy to do $20/3 so I just did $7 each and PP converted to EUR and GBP.

**Pull request closed without merging**